cas-server-webapp/src/main/java/org/jasig/cas/web/support/AbstractThrottledSubmissionHandlerInterceptorAdapter.java

```diff
@@ -0,0 +1,177 @@
+import org.springframework.http.HttpMethod;
+import org.springframework.web.servlet.handler.HandlerInterceptorAdapter;
+import org.springframework.web.servlet.ModelAndView;
+import org.springframework.webflow.execution.RequestContext;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.validation.constraints.Min;
+import javax.validation.constraints.NotNull;
+
+/**
+ * Abstract implementation of the handler that has all of the logic.
+ *
+ * @author Longze Chen
+ * @since 4.1.0
+ */
+public abstract class AbstractThrottledSubmissionHandlerInterceptorAdapter extends HandlerInterceptorAdapter implements InitializingBean {
+
+    private static final int DEFAULT_FAILURE_THRESHOLD = 3;
+
+    private static final int DEFAULT_FAILURE_RANGE_IN_SECONDS = 1;
+
+    private static final String DEFAULT_USERNAME_PARAMETER = "username";
+
+    /** The Logger instance. */
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Min(0)
+    private int failureThreshold = DEFAULT_FAILURE_THRESHOLD;
+
+    @Min(0)
+    private int failureRangeInSeconds = DEFAULT_FAILURE_RANGE_IN_SECONDS;
+
+    @NotNull
+    private String usernameParameter = DEFAULT_USERNAME_PARAMETER;
+
+    private double thresholdRate;
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
+        this.thresholdRate = (double) failureThreshold / (double) failureRangeInSeconds;
+    }
+
+    @Override
+    public final boolean preHandle(final HttpServletRequest request, final HttpServletResponse response, final Object o)
+            throws Exception {
+        if (!HttpMethod.POST.name().equals(request.getMethod())) {
+            return true;
+        }
+
+        if (exceedsThreshold(request)) {
+            recordThrottle(request);
+            request.setAttribute(WebUtils.CAS_ACCESS_DENIED_REASON, "screen.blocked.message");
+            response.sendError(HttpStatus.SC_FORBIDDEN,
+                    "Access Denied for user [" + request.getParameter(usernameParameter)
+                            + "] from IP Address [" + request.getRemoteAddr() + ']');
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public final void postHandle(
+            final HttpServletRequest request,
+            final HttpServletResponse response,
+            final Object o, final ModelAndView modelAndView
+    ) throws Exception {
+        if (!HttpMethod.POST.name().equals(request.getMethod())) {
+            return;
+        }
+
+        final RequestContext context = (RequestContext) request.getAttribute("flowRequestContext");
+
+        if (context == null || context.getCurrentEvent() == null) {
+            return;
+        }
+        final String handleErrorName = (String) context.getFlowScope().get("handleErrorName");
+
+        if (OpenScienceFrameworkAuthenticationExceptionHandler.triggerLoginThrottle(handleErrorName)) {
```

Comment:
To increase throttle, the request must have encountered one of the three dedicated exceptions in context.